File: backend/app/api/graphrag.py
from __future__ import annotations

from fastapi import (
    APIRouter,
    HTTPException,
)

from pydantic import (
    BaseModel,
    Field,
)

from graphrag.analyze_trend import (
    analyze_question_detail,
)


# ============================================================
# ROUTER
# ============================================================

router = APIRouter(
    prefix="/api/graphrag",
    tags=["GraphRAG"],
)


# ============================================================
# REQUEST
# ============================================================

class GraphRAGRequest(
    BaseModel
):
    question: str = Field(
        ...,
        min_length=2,
        description="패션 트렌드에 대한 질문",
        examples=[
            "스키니진은 다시 유행하고 있어?"
        ],
    )


# ============================================================
# HEALTH
# ============================================================

@router.get(
    "/health"
)
def graphrag_health():

    return {
        "status": "ok",
        "service": "graphrag",
    }


# ============================================================
# ANALYZE
# ============================================================

@router.post(
    "/analyze"
)
def analyze_trend(
    request: GraphRAGRequest,
):

    question = (
        request.question.strip()
    )


    try:

        result = (
            analyze_question_detail(
                question
            )
        )


        if not result[
            "success"
        ]:

            raise HTTPException(
                status_code=400,
                detail=result[
                    "message"
                ],
            )


        return result


    except HTTPException:
        raise


    except Exception as error:

        print(
            "GraphRAG API Error:",
            error,
        )


        raise HTTPException(
            status_code=500,
            detail=(
                "GraphRAG 분석 중 "
                "오류가 발생했습니다."
            ),
        )

File: frontend/src/api/graphrag.js
const API_BASE_URL = "http://127.0.0.1:8000";


export async function analyzeTrend(question) {
    const response = await fetch(
        `${API_BASE_URL}/api/graphrag/analyze`,
        {
            method: "POST",

            headers: {
                "Content-Type": "application/json",
            },

            body: JSON.stringify({
                question,
            }),
        }
    );


    if (!response.ok) {
        let detail =
            "GraphRAG 분석 중 오류가 발생했습니다.";

        try {
            const errorData =
                await response.json();

            detail =
                errorData.detail ||
                detail;

        } catch {
            // 응답이 JSON이 아니어도 기본 메시지 사용
        }


        const error =
            new Error(detail);

        error.status =
            response.status;

        throw error;
    }


    return await response.json();
}